data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/RouterAcknowledgementsRecordStrategy.java

```diff
@@ -0,0 +1,137 @@
+import org.opensearch.dataprepper.model.event.EventFactory;
+import org.opensearch.dataprepper.model.event.EventHandle;
+import org.opensearch.dataprepper.model.event.EventBuilder;
+import org.opensearch.dataprepper.model.event.EventMetadata;
+import org.opensearch.dataprepper.event.DefaultEventHandle;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
+
+public class RouterAcknowledgementsRecordStrategy implements RouterGetRecordStrategy {
+    private Set<Record> routedRecords;
+    private Set<Record> referencedRecords;
+    private AcknowledgementSetManager acknowledgementSetManager;
+    private EventFactory eventFactory;
+
+    public <C> RouterAcknowledgementsRecordStrategy(final EventFactory eventFactory, final AcknowledgementSetManager acknowledgementSetManager, final Collection<DataFlowComponent<C>> dataFlowComponents) {
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        this.eventFactory = eventFactory;
+        routedRecords = null;
+        referencedRecords = new HashSet<Record>();
+        /*
+         * If there are more than one sink and one of the sinks is
+         * pipeline connector, then we should make a copy of every
+         * record that is routed to more than one sink, so, to keep
+         * track of already routed records, initialize the set.
+         */
+        if (dataFlowComponents.size() > 1) {
+            for (DataFlowComponent<C> dataFlowComponent : dataFlowComponents) {
+                if (dataFlowComponent.getComponent() instanceof PipelineConnector) {
+                    routedRecords = new HashSet<Record>();
+                    break;
+                }
+            }
+        }
+    }
+
+    Set<Record> getReferencedRecords() {
+        return referencedRecords;
+    }
+
+    private void acquireEventReference(Record record) {
+        if (record.getData() == null) {
+            return;
+        }
+        if (referencedRecords.contains(record) || ((routedRecords != null) && routedRecords.contains(record))) {
+            EventHandle eventHandle = ((JacksonEvent)record.getData()).getEventHandle();
+            if (eventHandle != null) {
+                acknowledgementSetManager.acquireEventReference(eventHandle);
+            }
+        } else if (!referencedRecords.contains(record)) {
+            referencedRecords.add(record);
+        }
+    }
+    @Override
+    public Record getRecord(Record record) {
+        if (routedRecords == null) {
+            acquireEventReference(record);
+            return record;
+        }
+        if (!routedRecords.contains(record)) {
+            acquireEventReference(record);
+            routedRecords.add(record);
+            return record;
+        }
+        if (record.getData() instanceof JacksonSpan) {
```

Comment:
Is there a way that this strategy could use the `RouterCopyRecordsStrategy` so that this code is not duplicated?

If this is a large refactoring, I'm fine to tackle it later. But, right now, this important logic is duplicated between these two classes which can lead to development mistakes.

Reply:
I am thinking of removing RouterCopyRecordsStrategy completely some time later. I chose this option for now so that the impact of not using acknowledgements is minimal or even zero in the current release.

Reply:
It is easy to not duplicate the code and have just one file but each event that gets routed will at least have at least one check for `areAcknowledgementsEnabled`. Would that be Ok with you?

Reply:
I don't think it will be an expensive operation to check this boolean value.

Reply:
OK. Will make that change.